source/develop/release-management/features/virt/vm-lifecycle-with-kubevirt.md

```diff
@@ -0,0 +1,130 @@
+
+## Milestone 3: Dynamic Cluster State Managed by Kubevirt
+
+### Goal
+At this point we have basic monitoring of VMs. This milestone is about detecting more changes in running VMs in Kubevirt and report them to oVirt. Ideally, most of VmDynamic would be managed in Kubevirt and will be in oVirt only for display purpose. This allows us to switch VM management operations (shutdown, migrate, suspend, ..) to be done via Kubervirt.
+
+### Design
+1. Receive the VMs state from the node and updating the dynamic part in the VM representation.  
+2. Starting highly-available VMs in Kubevirt.
+3. Change VM management operations: hibernate, restore, resume, migrate, shutdown, reboot, cancel-migration, power-off.
+
+### Benefits
+1. Kubevirt becomes the owner of the runtime state of the cluster while oVirt manages its static state (host-deployments, VM templates, import VMs from external providers, and so on).  
+2. Many VM operations are delegated to Kubevirt.
+
+## Future Changes
+
+Future changes depend on having the knowledge of storage and network resources in Kubevirt and thus harder to predict.  
+
+## VM Representation
+
+I believe that the best way to represent the VM is in JSON as follow:  
+1. Scheduling data: the cluster-related data mentioned before, in the form of properties in JSON.  
+2. VM dynamic data: things like status of the VM, run_on_vds, console ip, and so on. Initially they should be set with default values. Once the monitoring in Kubevirt receives them from the node, it should update the VM representation and some component needs to listen to these changes and report them to the engine in oVirt. Note that we may want to report some of the dynamic data that exist today, like the applications list, in a different way.   
+3. VM configuration: the static configuration of the VM, including its devices.  
+
+In oVirt we are in the process of representing the VM configuration in XML (domxml) to eliminate the duplicate conversion of the data (once in the engine, converting the internal representation into a dictionary, and once in VDSM, converting the dictionary into domxml). We would prefer to keep this approach and minimize data conversions. In light of this approach, there are few options of how to represent the VM configuration:  
+1. To generate the VM representation in JSON. There is a proposal (on-going work?) to implement another binding for libvirt that will enable to pass the VM definition in JSON. With this, we would still have only one conversion of the VM configuration. That seems like the best approach since Kubernetes works with Yaml formats.  
+2. To have the VM configuration as XML blob inside JSON. This would mean that Kubevirt will have to parse VM-level data that it needs (for example, amount of memory for scheduling) by parsing the XML. The XML could be extended and place-holders can be replaced in Kubevirt but Kubevirt must not change other existing values in the XML.  
```

Comment:
The Kubevirt VM object has a metadata section where you can embed something, but kubevirt itself will not interact with that metadata. Also it will not parse xml.